internal/fs/buffer/inmemory_write_buffer.go

```diff
@@ -15,33 +15,156 @@
 package buffer
 
 import (
-	"bytes"
+	"fmt"
+
+	"github.com/googlecloudplatform/gcsfuse/internal/logger"
 )
 
+type offset struct {
+	// [ start, end) is the range of file offset for which the
+	// data is currently stored in the buffer.
+	start, end int64
+}
+
+type inMemoryBuffer struct {
+	buffer []byte
+	offset offset
+}
+
+// InMemoryWriteBuffer implements WriteBuffer and is stored in memory.
+// Refer to buffer.WriteBuffer for sample code usage.
 type InMemoryWriteBuffer struct {
-	// Holds the data to be written to GCS. At any time, Buffer holds the data for
-	// 2 GCS write calls.
-	buffer *bytes.Buffer
+	// Holds the incoming data from kernel write calls, currently being buffered
+	// for upload in the future.
+	current inMemoryBuffer
+
+	// Holds the data that is currently being uploaded asynchronously to GCS. Data
+	// in this buffer will persist until the entire data is successfully uploaded.
+	flushed inMemoryBuffer
+
+	// Holds the last offset to which data is written.
+	fileSize int64
+
+	// Holds the buffer's size (in bytes) to be allocated to currentBuffer and flushedBuffer.
+	bufferSize int64
 }
 
-// CreateInMemoryWriteBuffer creates a buffer with InMemoryWriteBuffer.buffer
-// set to nil. Memory is allocated to the buffer when the first write call comes.
-// This avoids unnecessarily bloating GCSFuse memory consumption.
-//
-// To allocate memory to the buffer, use WriteBuffer.InitializeBuffer.
-func CreateInMemoryWriteBuffer() *InMemoryWriteBuffer {
-	b := &InMemoryWriteBuffer{}
+// CreateInMemoryWriteBuffer returns a new InMemoryWriteBuffer.
+// Memory is lazily allocated to current and flushed buffer on their first usage.
+func CreateInMemoryWriteBuffer(bufferSizeMB uint) (*InMemoryWriteBuffer, error) {
+	if bufferSizeMB == 0 {
+		return nil, fmt.Errorf(ZeroSizeBufferError)
+	}
+	b := &InMemoryWriteBuffer{
+		bufferSize: int64(bufferSizeMB * MiB),
+	}
 	// TODO: set mtime attribute.
-	return b
+	logger.Debugf("TODO: set mtime attribute.")
+	return b, nil
 }
 
-func (b *InMemoryWriteBuffer) InitializeBuffer(sizeInMB int) {
-	ChunkSize = sizeInMB * MiB
-	if b.buffer == nil {
-		b.buffer = bytes.NewBuffer(make([]byte, 0, 2*sizeInMB*MiB))
+// Allocates memory to currentBuffer if it hasn't been allocated already.
+func (b *InMemoryWriteBuffer) ensureCurrentBuffer() error {
```

Comment:
The code for ensureCurrentBuffer and ensureFlushedBuffer is exactly same except for b.current vs. b.flushed. Better to merge them into a single function. This will be shorter calling code as well, and easier to test too.

E.g.

```golang
func (b *inMemoryBuffer) ensure(bufferSizeMB) error {
```

Reply:
shall I pass flushed or current buffer as a parameter then?

Reply:
@ashmeenkaur  Sorry for the late response. I somehow missed this earlier. 

Either way will do for now, as long as we reduce duplication of logic. We can come back and refactor later if ever needed.